lib/posix-vfs/vfs.c

```diff
@@ -0,0 +1,2013 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2025, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+#define _GNU_SOURCE
+
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/time.h>
+#include <sys/vfs.h>
+
+#include <uk/alloc.h>
+#include <uk/atomic.h>
+#include <uk/errptr.h>
+#include <uk/essentials.h>
+#include <uk/fs.h>
+#include <uk/fs/driver.h>
+#include <uk/fs/pathutil.h>
+#include <uk/mutex.h>
+#include <uk/posix-fdio.h>
+#include <uk/posix-vfs.h>
+#include <uk/posix-vfsroot.h>
+#include <uk/spinlock.h>
+
+#if CONFIG_LIBPOSIX_VFS_MULTICTX
+#include <uk/init.h>
+#include <uk/refcount.h>
+#include <uk/thread.h>
+#endif /* CONFIG_LIBPOSIX_VFS_MULTICTX */
+
+#if CONFIG_LIBPOSIX_PROCESS_CLONE
+#include <uk/process.h>
+#endif /* CONFIG_LIBPOSIX_PROCESS_CLONE */
+
+#if CONFIG_LIBPOSIX_VFS_MOUNT_CLEANUP
+#include <uk/init.h>
+#endif /* CONFIG_LIBPOSIX_VFS_MOUNT_CLEANUP */
+
+/* posix-ey perms checking */
+#if CONFIG_LIBPOSIX_VFS_PERMISSIONS
+int uk_fs_checkperm(int desired, int mode, int uid, int gid)
+{
+	/* TODO: impl & integrate w/ posix-user */
+	return 0;
+}
+#endif /* CONFIG_LIBPOSIX_VFS_PERMISSIONS */
+
+/* mount table */
+struct vfs_mtab_entry {
+	const struct uk_file *point;
+	const struct uk_file *root;
+	const struct uk_file *parent_root;
+};
+
+struct vfs_mtab {
+	struct uk_mutex lock;
+	size_t top;
+	struct vfs_mtab_entry tab[CONFIG_LIBPOSIX_VFS_MAX_MOUNTS];
+};
+
+static struct vfs_mtab mtab = {
+	.lock = UK_MUTEX_INITIALIZER(mtab.lock),
+	.top = 0
+};
+
+#if CONFIG_LIBPOSIX_VFS_MOUNT_CLEANUP
+
+static
+void vfs_mtab_cleanup(struct uk_term_ctx *tctx __unused)
```

Comment:
I guess we don't need to protect this with mtab lock cause it is during init? If so, can you add a comment?